internal/cachecontroller/cache_controller.go

```diff
@@ -87,15 +88,23 @@ func (c *InMemoryCacheController) DetermineInvalidation(
 		return entry.LastModified
 	}
 
-	lastModified, err, _ := c.sf.Do(storeID, func() (interface{}, error) {
-		return c.findChangesAndInvalidate(ctx, storeID)
-	})
-
-	if err != nil {
-		return time.Time{}
+	_, present := c.mutexMap[storeID]
+	if !present {
```

Comment:
Why not create the entry with the lock already acquired? So that when you prompt, if it exists you basically exit immediately. AKA, reducing the number of accesses to the map by one since the creator of the lock will also, be the owner of it.

Reply:
found a bigger bug - the mutexMap itself entry is not protected.  Switch to have a single mutex protecting the map[string]struct{} instead